openhtf/io/frontend/__main__.py

```diff
@@ -32,97 +32,223 @@
+      framework: A dictionary representation of a TestExecutor object.
+      test: A dictionary representation of a TestState object.
+    """
+
+
+class StationStore(threading.Thread):
+  """Self-updating store of stations visible on the local network(s).
+
+  Station data is stored in the 'stations' attribute, a dictionary mapping
+  tuples of (host_ip, port) to Station records.
+  """
+  def __init__(self):
+    super(StationStore, self).__init__()
+    self._live = False
+    self.hostname = socket.gethostname()
+    self.stations = {}
+
+  @staticmethod
+  def MessageFramework(host, port, method='GET', message=None):
+    """Interact with a running instance of openhtf via its HTTP API.
+
+    The default method, 'GET', returns a JSON representation of station state,
+    while 'POST' can be used to respond to prompts. For example:
+
+      # Get the state of the openhtf instance at 192.168.2.25:10500.
+      station_state = MessageFramework(192.168.2.25, 10500)
+
+      # Respond to a prompt for the DUT's serial number:
+      MessageFramework(192.168.2.25, 10500, method='POST', 'SN-NCC1701D')
+
+    Args:
+      host: The IP address of the host running openhtf.
+      port: The TCP port of openhtf's HTTP API on host.
+      method: 'GET' or 'POST'.
+      message: Prompt response for POSTs. Ignored for GETs.
+    """
+    result = None
+    conn = httplib.HTTPConnection(host, int(port))
+    args = [method, '/']
+    if message and method == 'GET':
+      _LOG.warning("Message provided for 'GET' request will be ignored: %s",
+                   message)
+    if message and method == 'POST':
+      args.append(message)
+    try:
+      conn.request(*args)
+      if method == 'GET':
+        response = conn.getresponse()
+        if response.status != 200:
+          result = None
+        else:
+          result = json.loads(response.read())
+    except socket.error as e:
+      result = None
+    conn.close()
+    return result
+
+  def _Discover(self):
+    """Use multicast to discover stations on the local network."""
+    responses = multicast.send(multicast.PING_STRING,
+                               FLAGS.discovery_address,
+                               FLAGS.discovery_port,
+                               FLAGS.discovery_ttl)
+    for host, response in responses:
+      try:
+        port = json.loads(response)[multicast.PING_RESPONSE_KEY]
+        self._Track(host, port)
+      except (KeyError, ValueError):
+        _LOG.warning('Ignoring unrecognized discovery response from %s: %s' % (
+            host, response))
+
+  def run(self):
+    """Continuously scan for new stations and add them to the store."""
+    self._live = True
+    while self._live:
+      self._Discover()
+      time.sleep(FLAGS.discovery_interval_s)
+
+  def _Track(self, host, port):
+    """Start tracking the given station."""
+    station = self.stations.setdefault((host, port), Station())
+    if station.station_id is None:
```

Comment:
It won't ever be None because it's UNKNOWN_STATION_ID by default. You can put None there instead to get that behavior here